cores/esp8266/mmu_iram.h

```diff
@@ -127,73 +127,66 @@ bool mmu_is_icache(const void *addr) {
 static inline __attribute__((always_inline))
 uint8_t mmu_get_uint8(const void *p8) {
   ASSERT_RANGE_TEST_READ(p8);
-  uint32_t val = (*(uint32_t *)((uintptr_t)p8 & ~0x3));
-  uint32_t pos = ((uintptr_t)p8 & 0x3) * 8;
-  val >>= pos;
-  return (uint8_t)val;
+  return *(uint32_t *)((uintptr_t)p8 & ~0x3) >> (uintptr_t)p8 * 8;
 }
 
 static inline __attribute__((always_inline))
 uint16_t mmu_get_uint16(const uint16_t *p16) {
   ASSERT_RANGE_TEST_READ(p16);
-  uint32_t val = (*(uint32_t *)((uintptr_t)p16 & ~0x3));
-  uint32_t pos = ((uintptr_t)p16 & 0x3) * 8;
-  val >>= pos;
-  return (uint16_t)val;
+  return *(uint32_t *)((uintptr_t)p16 & ~0x3) >> (uintptr_t)p16 * 8;
 }
 
 static inline __attribute__((always_inline))
 int16_t mmu_get_int16(const int16_t *p16) {
   ASSERT_RANGE_TEST_READ(p16);
-  uint32_t val = (*(uint32_t *)((uintptr_t)p16 & ~0x3));
-  uint32_t pos = ((uintptr_t)p16 & 0x3) * 8;
-  val >>= pos;
-  return (int16_t)val;
+  return *(uint32_t *)((uintptr_t)p16 & ~0x3) >> (uintptr_t)p16 * 8;
 }
 
 static inline __attribute__((always_inline))
 uint8_t mmu_set_uint8(void *p8, const uint8_t val) {
   ASSERT_RANGE_TEST_WRITE(p8);
-  uint32_t pos = ((uintptr_t)p8 & 0x3) * 8;
-  uint32_t sval = val << pos;
-  uint32_t valmask =  0x0FF << pos;
-
   uint32_t *p32 = (uint32_t *)((uintptr_t)p8 & ~0x3);
   uint32_t ival = *p32;
-  ival &= (~valmask);
-  ival |= sval;
-  *p32 = ival;
+  uint32_t valmask, sval;
+
+  __asm__ ("ssa8b\t%2\n\t"
+           "src\t%0, %3, %3\n\t"
+           "src\t%1, %4, %4"
+           : "=&a"(valmask), "=a"(sval)
+           : "r"(p8), "r"(~0xFF), "r"(val), "r"(ival));
```

Comment:
I don't see `ival` being used here, it would be `%5`.

Reply:
> I don't see `ival` being used here, it would be `%5`.

`"r"(ival)` is used only for injecting dependency, in order to ensure `ival = *p32;` precedes the asm block.
without this, it may be placed between the asm block and `*p32 = (ival & valmask) | sval;`, thus will cause pipeline interlock stall of 1 cycle. (see Xtensa ISA refman. p608)

Reply:
sorry for quite low-level machine-dependent coding ;)
added comments for the explanation.

Reply:
I wonder if this is necessary. The compiler already appears to be pipeline aware, and often preloads values before they are needed. And, since your Extended ASM block does not use the "memory" clobber parameter. The compiler would have no need to reload the register after the ASM block.

Reply:
good news: after 3498137c37ba06baf06c9823b1755701b4b803d4 (or 51bcd52bc8f99b310142cd3ba65d7a6934b472c9), such a stall seem not to be seen if without `"r"(ival)`:
```
void test(void* p, char c) {
    mmu_set_uint8(p, c);
}
```
```
	.file	"test.c"
	.text
	.literal_position
	.align	4
	.global	test
	.type	test, @function
test:
	movi.n	a4, -4
	and	a4, a2, a4
	l32i.n	a5, a4, 0
	movi	a6, -0x100
#APP
# 9 "test.c" 1
	ssa8b	a2
	src	a7, a6, a6
	src	a2, a3, a3
# 0 "" 2
#NO_APP
	and	a5, a5, a7
	or	a2, a2, a5
	s32i.n	a2, a4, 0
	ret.n
	.size	test, .-test
	.ident	"GCC: (GNU) 10.2.0"
```
however, `mmu_set_*()` are inline functions; their function bodies will be meld into another, thus there may be exist other contexts of processing that interfere with the code arrangement.
clobbering "memory" also works well but overdoing; that is non-local read/write memory barrier, so presses the other contexts of processing for non-urgent store-to-memory operations, and puts obstacles in the way of the compiler's optimizer.

in conclusion, i am sure that it is appropriate the use of unused input operand `"r"(ival)` in asm block, because it operates as read-only memory barrier on only a specific target.

Reply:
LGTM